components/proxy/src/main/java/com/hotels/styx/startup/ProxyServerSetUp.java

```diff
@@ -39,32 +48,51 @@ public ProxyServerSetUp(PipelineFactory pipelineFactory) {
         this.pipelineFactory = requireNonNull(pipelineFactory);
     }
 
-    public HttpServer createProxyServer(StyxServerComponents config) {
-        HttpHandler pipeline = pipelineFactory.create(config);
+    public HttpServer createProxyServer(StyxServerComponents components) {
+        notifyProxyStatus(
+                components.environment().configStore(),
+                components.environment().configuration().proxyServerConfig());
 
-        HttpServer proxyServer = new ProxyServerBuilder(config.environment())
-                .httpHandler(pipeline)
-                .onStartup(() -> initialisePlugins(config.plugins()))
-                .build();
+        // TODO see https://github.com/HotelsDotCom/styx/issues/382
 
-        proxyServer.addListener(new PluginsNotifierOfProxyState(config.plugins()), sameThreadExecutor());
-        return proxyServer;
+        while (!arePluginsLoaded(components)) {
+            sleep(100, MILLISECONDS);
+        }
+
+        List<NamedPlugin> plugins = components.environment().configStore()
+                .valuesStartingWith("plugins", NamedPlugin.class);
+
+        return createProxyServer0(components, plugins);
     }
 
-    private static void initialisePlugins(Iterable<NamedPlugin> plugins) {
-        int exceptions = 0;
+    private static boolean arePluginsLoaded(StyxServerComponents components) {
+        PluginPipelineStatus status = components.environment().configStore()
+                .get(PLUGIN_PIPELINE_STATUS_KEY, PluginPipelineStatus.class)
+                .orElse(INCOMPLETE);
 
-        for (NamedPlugin plugin : plugins) {
-            try {
-                plugin.styxStarting();
-            } catch (Exception e) {
-                exceptions++;
-                LOG.error("Error starting plugin '{}'", plugin.name(), e);
-            }
+        if(status == AT_LEAST_ONE_PLUGIN_FAILED) {
+            throw new IllegalStateException("One or more plugins failed to start");
         }
 
-        if (exceptions > 0) {
-            throw new RuntimeException(format("%s plugins failed to start", exceptions));
+        return status == ALL_PLUGINS_COMPLETE;
+    }
+
+    private HttpServer createProxyServer0(StyxServerComponents components, List<NamedPlugin> plugins) {
+        HttpHandler pipeline = pipelineFactory.create(components, plugins);
+
+        HttpServer proxyServer = new ProxyServerBuilder(components.environment())
+                .httpHandler(pipeline)
+                .build();
+
+        proxyServer.addListener(new PluginsNotifierOfProxyState(plugins), sameThreadExecutor());
+        return proxyServer;
+    }
+
+    private static void sleep(long duration, TimeUnit unit) {
+        try {
+            Thread.sleep(unit.toMillis(duration));
```

Comment:
As per Java best practices, at least set the interrupted status of the Thread when an InterruptedException cannot be re-thrown.

I would use function composition to deal with these exceptions (similar to your throwsException() function) instead of declaring a sleep function.